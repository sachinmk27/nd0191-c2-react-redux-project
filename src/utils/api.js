import {
  _getQuestions,
  _getUsers,
  _saveQuestion,
  _saveQuestionAnswer,
} from './_DATA';

export function getInitialData() {
  return Promise.all([_getQuestions(), _getUsers()]).then(
    ([questions, users]) => ({
      questions,
      users,
    })
  );
}

export function saveQuestion(info) {
  return _saveQuestion(info);
}

export function saveQuestionAnswer(info) {
  return _saveQuestionAnswer(info);
}
