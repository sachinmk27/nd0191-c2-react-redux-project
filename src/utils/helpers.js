export function formatQuestion(question, author, authedUser) {
  const total =
    question.optionOne.votes.length + question.optionTwo.votes.length;
  return {
    ...question,
    authorName: author.name,
    authorAvatarURL: author.avatarURL,
    hasAnswered: !!authedUser.answers[question.id],
    hasAnsweredOption: authedUser.answers[question.id],
    noOfResponses: {
      total,
      optionOne: Math.round((question.optionOne.votes.length * 100) / total),
      optionTwo: Math.round((question.optionTwo.votes.length * 100) / total),
    },
  };
}

export function formatDate(timestamp) {
  const d = new Date(timestamp);
  const time = d.toLocaleTimeString('en-US');
  return time.substr(0, 5) + time.slice(-2) + ' | ' + d.toLocaleDateString();
}
