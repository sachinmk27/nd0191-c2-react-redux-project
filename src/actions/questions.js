import { hideLoading, showLoading } from 'react-redux-loading-bar';
import { saveQuestion, saveQuestionAnswer } from '../utils/api';

export const RECEIVE_QUESTIONS = 'RECEIVE_QUESTIONS';
export const SET_QUESTION_ANSWER = 'SET_QUESTION_ANSWER';
export const ADD_QUESTION = 'ADD_QUESTION';

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function setQuestionAnswer({ authedUser, qid, answer }) {
  return {
    type: SET_QUESTION_ANSWER,
    authedUser,
    qid,
    answer,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleSetQuestionAnswer(qid, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    dispatch(showLoading());
    saveQuestionAnswer({ authedUser, qid, answer })
      .then(() => {
        dispatch(setQuestionAnswer({ authedUser, qid, answer }));
      })
      .then(() => {
        dispatch(hideLoading());
      })
      .catch((e) => {
        console.warn('Error in handleSetQuestionAnswer: ', e);
        alert('There was an error saving the answer. Try again.');
      });
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authedUser: author } = getState();
    dispatch(showLoading());
    saveQuestion({ optionOneText, optionTwoText, author })
      .then((question) => {
        dispatch(addQuestion(question));
      })
      .then(() => {
        dispatch(hideLoading());
      })
      .catch((e) => {
        console.warn('Error in handleAddQuestion: ', e);
        alert('There was an error saving the question. Try again.');
      });
  };
}
