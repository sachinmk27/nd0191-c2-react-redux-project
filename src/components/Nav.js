import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setAuthedUser } from '../actions/authedUsers';

const Nav = ({ authedUser, dispatch }) => {
  const handleLogout = () => {
    dispatch(setAuthedUser(null));
  };
  return (
    <div className="navbar bg-primary text-base-100">
      <nav className="flex-1">
        <ul className="flex gap-2">
          <li>
            <Link className="btn btn-link text-base-100 no-underline" to="/">
              Home
            </Link>
          </li>
          <li>
            <Link
              className="btn btn-link text-base-100 no-underline"
              to="/leaderboard"
            >
              Leaderboard
            </Link>
          </li>
          <li>
            <Link className="btn btn-link text-base-100 no-underline" to="/add">
              New
            </Link>
          </li>
        </ul>
      </nav>
      <div className="flex gap-4">
        <div className="avatar">
          <div className="w-10 rounded-full ring ring-primary ring-offset-base-100 ring-offset-2">
            <img src={authedUser.avatarURL} alt={authedUser.name} />
          </div>
        </div>
        <p>{authedUser.name}</p>
        <button onClick={handleLogout} className="btn btn-sm bg-base-100">
          Logout
        </button>
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }) => {
  return {
    authedUser: users[authedUser],
  };
};

export default connect(mapStateToProps)(Nav);
