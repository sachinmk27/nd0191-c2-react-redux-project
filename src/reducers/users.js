import { SET_QUESTION_ANSWER } from '../actions/questions';
import { RECEIVE_USERS } from '../actions/users';

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case SET_QUESTION_ANSWER:
      const { authedUser, qid, answer } = action;
      return {
        ...state,
        [authedUser]: {
          ...state[authedUser],
          answers: {
            ...state[authedUser].answers,
            [qid]: answer,
          },
        },
      };
    default:
      return state;
  }
}
