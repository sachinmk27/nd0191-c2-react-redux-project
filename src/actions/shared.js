import { hideLoading, showLoading } from 'react-redux-loading-bar';
import { getInitialData } from '../utils/api';
import { receiveQuestions } from './questions';
import { receiveUsers } from './users';

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return getInitialData().then(({ questions, users }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(hideLoading());
    });
  };
}
