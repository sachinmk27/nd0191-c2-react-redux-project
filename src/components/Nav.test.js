import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import Nav from './Nav';
import { Provider } from 'react-redux';
import { combineReducers, createStore } from 'redux';
import authedUser from '../reducers/authedUser';
import users from '../reducers/users';
import { setAuthedUser } from '../actions/authedUsers';
import { _getUsers } from '../utils/_DATA';
import { receiveUsers } from '../actions/users';

describe('Nav', () => {
  test('snapshot', async () => {
    const store = createStore(combineReducers({ authedUser, users }));
    const usersData = await _getUsers();
    store.dispatch(receiveUsers(usersData));
    store.dispatch(setAuthedUser(Object.keys(usersData)[0]));

    const view = render(
      <Provider store={store}>
        <MemoryRouter>
          <Nav />
        </MemoryRouter>
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });

  test('should have links', async () => {
    const store = createStore(combineReducers({ authedUser, users }));
    const usersData = await _getUsers();
    store.dispatch(receiveUsers(usersData));
    store.dispatch(setAuthedUser(Object.keys(usersData)[0]));

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Nav />
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByText('Leaderboard')).toBeInTheDocument();
    expect(screen.getByText('New')).toBeInTheDocument();
    expect(screen.getByText('Logout')).toBeInTheDocument();
  });
});
