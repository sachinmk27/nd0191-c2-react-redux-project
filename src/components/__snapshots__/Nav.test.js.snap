// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="navbar bg-primary text-base-100"
      >
        <nav
          class="flex-1"
        >
          <ul
            class="flex gap-2"
          >
            <li>
              <a
                class="btn btn-link text-base-100 no-underline"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                class="btn btn-link text-base-100 no-underline"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li>
              <a
                class="btn btn-link text-base-100 no-underline"
                href="/add"
              >
                New
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="flex gap-4"
        >
          <div
            class="avatar"
          >
            <div
              class="w-10 rounded-full ring ring-primary ring-offset-base-100 ring-offset-2"
            >
              <img
                alt="Sarah Edo"
                src="sarahedo.jpg"
              />
            </div>
          </div>
          <p>
            Sarah Edo
          </p>
          <button
            class="btn btn-sm bg-base-100"
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="navbar bg-primary text-base-100"
    >
      <nav
        class="flex-1"
      >
        <ul
          class="flex gap-2"
        >
          <li>
            <a
              class="btn btn-link text-base-100 no-underline"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="btn btn-link text-base-100 no-underline"
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li>
            <a
              class="btn btn-link text-base-100 no-underline"
              href="/add"
            >
              New
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="flex gap-4"
      >
        <div
          class="avatar"
        >
          <div
            class="w-10 rounded-full ring ring-primary ring-offset-base-100 ring-offset-2"
          >
            <img
              alt="Sarah Edo"
              src="sarahedo.jpg"
            />
          </div>
        </div>
        <p>
          Sarah Edo
        </p>
        <button
          class="btn btn-sm bg-base-100"
        >
          Logout
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
